File: Daily/20131002/mcpc2012/mcpc2012/markers/markers.cpp
/*
  markers.cpp
  The Mark of a Wizard, MCPC 2012 Problem E
  C++ version by Michael Goldwasser
*/

#include <iostream>
#include <fstream>
#include <vector>
#include <set>
using namespace std;

#define MAX_TRIALS 16
#define MAX_EDGES 35
#define MAX_SITES 17
#define MAX_DEGREE 6
#define MAX_PATH 7
void validate(int);     // used only for judging

ifstream fin("markers.in");

int degree[MAX_SITES];
int adj[MAX_SITES][MAX_DEGREE];    // adj[v][j] is the index of the jth neighbor of v
int wgt[MAX_SITES][MAX_DEGREE];    // wgt[v][j] is the distance of edge connecting v to its jth neighbor
int topo[MAX_SITES];
int sp[MAX_SITES];
int n;

int toInt(char c) {
    int k = (c - 'A');
    int answer = (k >= 0 && k <= 25 ? k : 26 + (c-'a'));
    if (answer < 0 || answer >= n)
	cerr << "PROBLEM with neighbor label " << c << endl;
    return answer;
}

void computeTopo() {
    int indeg[n];
    int k(0);
    vector<int> ready;
    for (int v=0; v<n; v++) {
	indeg[v] = 0;
    }

    for (int v=0; v<n; v++)
	for (int j=0; j<degree[v]; j++)
	    indeg[adj[v][j]]++;

    for (int v=0; v<n; v++) {
	if (indeg[v] == 0)
	    ready.push_back(v);
    }
    while (!ready.empty()) {
	int u = ready[ready.size()-1];
	ready.pop_back();
	topo[k++] = u;
	for (int j=0; j<degree[u]; j++) {
	    if (--indeg[adj[u][j]] == 0)
		ready.push_back(adj[u][j]);
	}
    }
    validate(k);   // for judging use only
}

bool isSolution(const set<int>& marked) {
    set<int> good;
    good.insert(n-1);
    for (int k=n-2; k >= 0; k--) {
	int u = topo[k];
	int countgood(0);
	for (int j=0; j<degree[u]; j++) {
	    if (good.count(adj[u][j]) == 1 && sp[u] == wgt[u][j] + sp[adj[u][j]])
		countgood++;
	}
	if (countgood >= 1 && (marked.count(u) == 1 || countgood == degree[u]))
	    good.insert(u);
    }
    return good.count(0) == 1;
}

int solve(set<int>& marked, int next) {
    if (isSolution(marked))
	return marked.size();
    int best = MAX_PATH;                // this must exist
    if (marked.size() < MAX_PATH - 1)
	for (int j=next; j<n; j++) {
	    marked.insert(j);
	    best = min(best, solve(marked, j+1));
	    marked.erase(j);
	}
    return best;
}

int main() {
    int trial(0);
    while (true) {
	fin >> n;
	if (n==0) break;
	if (++trial > MAX_TRIALS)
	    cerr << "PROBLEM with too many data sets" << endl;

	// read input
	for (int i=0; i<n; i++) {
	    char c;
	    fin >> c;
	    int v = toInt(c);
	    fin >> degree[v];
	    for (int j=0; j < degree[v]; j++) {
		fin >> c;
		adj[v][j] = toInt(c);
		fin >> wgt[v][j];
	    }
	}

	// Step 1) compute topo order of DAG
	computeTopo();

	// Step 2) compute shortest path lengths from all vertices to sink
	sp[n-1] = 0;   // the sink
	for (int k=n-2; k>=0; k--) {
	    int u = topo[k];
	    sp[u] = wgt[u][0] + sp[adj[u][0]];
	    for (int j=1; j<degree[u]; j++)
		sp[u] = min(sp[u], wgt[u][j] + sp[adj[u][j]]);
	}
	cout << sp[0];  // print shortest path distance (but no newline)

	// Step 3) try to find solution with at most MAX_PATH - 1 markers
	// (otherwise, certainly exists solution with MAX_PATH markers)
	set<int> marked;
	cout << " " << solve(marked, 0) << endl;
    }
}



// used only for judges
void validate(int k) {
    if (n > MAX_SITES)
	cerr << "PROBLEM with too many vertices" << endl;
    if (k != n)
	cerr << "PROBLEM computing topo" << endl;
    if (topo[n-1] != n-1)
	cerr << "PROBLEM as sink is not unique" << endl;
    int e(0);
    for (int j=0; j<n; j++) {
	e += degree[j];
	if (degree[j] > MAX_DEGREE)
	    cerr << "PROBLEM with degree of vertex " << j << endl;
    }
    if (e > MAX_EDGES)
	cerr << "PROBLEM with too many total edges" << endl;

}

File: Daily/20131002/mcpc2012/mcpc2012/markers/markers.java
/*
  markers.java
  The Mark of a Wizard, MCPC 2012 Problem E
  Java version by Andy Harrington
*/

/* 
Get graph.  
Find min dists by memoizing
Remove all edges not minimal (then no need for lengths any more)
Any node with edge from it removed would need a marker if the node is used.
Recursively mark vertices still in possible use via memoization
Mark others as having large dist
Sort by dist to get reverse topological order
Generate all subsets of marks with few enough marks,
and test each to see if sufficient via reverse topological check
if sufficient, reduce minMarks

Since there is a minimal length path with at most 7 tunnels, there
exists a solution with 7 or less markers. That bounds the number of
possible marker subsets to consider to C(17, 7) + C(17,6) + ... + C(17,0) = 41226.
*/

import java.util.*;
import java.io.*;
import static java.lang.Math.*;

class Node implements Comparable<Node> {
   List<Node> next = new ArrayList<Node>(); // posssible next nodes
   int[] wt;
   char name;
   boolean needsMark, canUse, doMark;
      // nextMark[i] is next mark after edge from this to next[i]
   int dist = -1;
   
   void init(Scanner in, Node[] nodes) 
   {
      String seq="ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghij";
      name = in.next().charAt(0);
      int n = in.nextInt();
      wt = new int[n];
      for (int i = 0; i < n; i++) {
         next.add(nodes[seq.indexOf(in.next())]);
         wt[i] = in.nextInt();
      }
   }
   
   public int compareTo(Node nd) // allows sort by distance
   {
      return dist - nd.dist;
   }
   
   int getDist()  // basic memoized minimum path length algorithm
   {
      if (dist == -1) {
         dist = 100000;
         int i = 0;
         for (Node nd : next) {
            dist = min(dist, wt[i] + nd.getDist());
            i++;
         }
      }  
      return dist;
   }
   
   void killLong()  //remove edges that are too long
   {
      int k = next.size();
      for (int i = k-1;  i >= 0; i --) // backward allows use of original wt
         if (dist  != wt[i]  + next.get(i).dist) {
            markers.prp(" " + name + next.get(i).name);
            next.remove(i);
         }
      needsMark = next.size() < k;              
   }
   
   void setReachable()  // seach forward for nodes reachable after edge removal
   {
      if (canUse) return;
      canUse = true;
      for (Node nd : next) 
         nd.setReachable();
   }
   
   void setCanUse()  // reverse topological check if node is useable
                     // with current marker choices
   {
      if (! doMark && needsMark) 
         canUse = false;
      else {
         int okTunnels = 0;
         for (Node nd : next)
            if (nd.canUse)
               okTunnels++;
         canUse = okTunnels == next.size()  || okTunnels > 0 && doMark; 
      }
   }

   public String toString()
   {  
      String after = "";
      for (Node nd : next) 
         after += nd.name;
      return String.format("%s: %s, d: %s; nM:%s;  cU: %s",
                              name, after, dist, needsMark, canUse);    
   }
}
   
public class markers {
   static int MAX_STEPS = 7, MAX_EDGES=35;
   static Node[] nodes;  
   static int tot;  // intial count of nodes; later drop unreachable
   static int minMarks;
    
   public static void main(String[] args) throws Exception {
      Scanner in = new Scanner(new File("markers.in"));
      tot = in.nextInt();
      while (tot > 0) {
         nodes = new Node[tot];
         for (int i =0; i < tot; i ++)
            nodes[i] = new Node(); // uninitialized, but can reference
         int totEdges = 0;
         for (Node nd : nodes) {
            nd.init(in, nodes);   // input data initializes node
            totEdges += nd.next.size();
         }
         if (totEdges > MAX_EDGES)
            pr("Too many edges! " + totEdges);
         nodes[tot-1].dist = 0;  // at destination
         int totDist = nodes[0].getDist(); //recursive memoize, find dist to end
         pr("dump edges: ");
         for (Node nd : nodes)
            nd.killLong(); // dump nonoptimal edges
         nodes[tot-1].canUse = true;
         nodes[0].setReachable(); // label all now reachable (canUse) from start
         for (Node nd : nodes) {
            if (!nd.canUse) {
                  nd.dist = 100000;  // will be sorted out of the way
                  tot--;             // one less active Node
            }
            for (Node nxt : nd.next) 
               if (!nd.canUse || !nxt.canUse)
                  prp(" " + nd.name + nxt.name); 
         }
         pr("\nGood edges:");
         for (Node nd : nodes) 
            if (nd.canUse) {
               for (Node nxt : nd.next) 
                  prp(" " + nd.name + nxt.name); 
            }
         pr("");   
         Arrays.sort(nodes); // now reverse topogogical order of tot nodes
//         for (Node nd : nodes)
//            pr(""+nd);
         minMarks = min(MAX_STEPS, tot-1); // could mark a single whole path
         pr(tot + " " + minMarks);
         pr("Marked edge sets:");
         doSubsets(1, 0);
         System.out.println(totDist + " " + minMarks);  
         tot = in.nextInt();
      }
   }
   
   static void doSubsets(int nextMark, int marks) {
      if (marks >= minMarks) return;
      for (int i = 1; i < tot; i++) // see if marks sufficient
         nodes[i].setCanUse();      // set in reverse topo order
      if (nodes[tot-1].canUse) {    // reach starting node?
         minMarks = marks;
         for (int i = 0; i < tot; i++) 
            if (nodes[i].doMark) {
               for (Node nd : nodes[i].next)
                  if (nd.canUse)
                     prp(" " + nodes[i].name + nd.name);
            }
            pr(": " + marks);
         return;
      }
      for (int i = nextMark; i < tot; i++) { // add one more mark
         nodes[i].doMark = true;
         doSubsets(i+1, marks+1);
         nodes[i].doMark = false;
      }
   }
         
   // only judge checks follow
   static boolean DEBUG = true;
    
    
   static void prp(String msg) {
      if (DEBUG) System.err.print(msg);
   }
    
   static void pr(String msg) {
      prp(msg+"\n");
   }
}
